### PR TITLE
Adds the opt_einsum package

### PR DESCRIPTION
**Pull request by dgasmith**
This package extends the `np.einsum` function to both take advantage of contraction path optimization and BLAS utilization, i.e., it makes `np.einsum` really fast.

Upstream can be found [here](https://github.com/dgasmith/opt_einsum).

### CODE REVIEW

File: recipes/opt_einsum/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    - py.test -v ./test
+
+about:
+  home: http://github.com/dgasmith/opt_einsum
+  license: MIT
+  license_family: MIT
+  license_file: LICENSE
+  summary: 'A contraction optimizer for the NumPy Einsum function.'
+
+  description: >
```
Comment (synapticarbors) [inline recipes/opt_einsum/meta.yaml]: Typically the `description` is not this long or detailed.

**Comment by dgasmith** [comment]
@synapticarbors Thanks for the review, which version in `setup.py` is out of date? They all like v1.0.1 to me.

File: recipes/opt_einsum/meta.yaml (hunk 1 above)
Comment (dgasmith) [inline recipes/opt_einsum/meta.yaml]: Is this ok? We debated a bit on how long the description could be.

File: recipes/opt_einsum/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    1000 loops, best of 3: 324 µs per loop
+    ```
+    
+    The above will automatically find the optimal contraction order, in this case identical to that of the optimized function above, and compute the products for you. In this case, it even uses `np.dot` under the hood to exploit any vendor BLAS functionality that your NumPy build has!
+  dev_url: https://github.com/dgasmith/opt_einsum
+
+extra:
+  recipe-maintainers:
+    - dgasmith
+    - loriab
```
Comment (loriab) [inline recipes/opt_einsum/meta.yaml]: Yes, I'm ok being added as a maintainer --- we had discussed this offline. Thanks for asking!

**Pull request merged**